src/AutomaticScheduling.cpp

```diff
@@ -0,0 +1,268 @@
+        map<string, vector<Function> >::const_iterator I = call_to_callee.find(f.name());
+        if (I == call_to_callee.end()) {
+            call_to_callee[f.name()] = vector<Function>();
+            I = call_to_callee.find(f.name());
+        }
+        return I->second;
+    }
+
+    /** Return list of functions transitively called by f. This does
+     * not include f. */
+    vector<Function> transitive_calls(Function f) {
+        const vector<Function> &dir_calls = calls(f);
+        set<string> visited;
+        vector<Function> result(dir_calls.begin(), dir_calls.end());
+        visited.insert(f.name());
+        for (vector<Function>::const_iterator I = dir_calls.begin(), E = dir_calls.end(); I != E; ++I) {
+            transitive_calls_helper(*I, result, visited);
+        }
+        return result;
+    }
+private:
+    map<string, Function> name_to_func;
+    map<string, vector<Function> > call_to_callee;
+    map<string, vector<Function> > call_to_caller;
+
+    void construct(Function f, set<string> &visited) {
+        visited.insert(f.name());
+        map<string, Function> calls = find_direct_calls(f);
+        for (map<string, Function>::iterator I = calls.begin(), E = calls.end(); I != E; ++I) {
+            string name = I->first;
+            Function func = I->second;
+            name_to_func[name] = func;
+            call_to_callee[f.name()].push_back(func);
+            call_to_caller[func.name()].push_back(f);
+
+            if (visited.find(I->first) == visited.end()) {
+                construct(I->second, visited);
+            }
+        }
+    }
+
+    void transitive_calls_helper(Function f, vector<Function> &result, set<string> &visited) {
+        visited.insert(f.name());
+        const vector<Function> &dir_calls = calls(f);
+        for (vector<Function>::const_iterator I = dir_calls.begin(), E = dir_calls.end(); I != E; ++I) {
+            Function call = *I;
+            if (visited.find(call.name()) == visited.end()) {
+                result.push_back(call);
+                transitive_calls_helper(call, result, visited);
+            }
+        }
+    }
+};
+
+/** Recursively reset the schedules for the given function and all of
+ * the functions it calls. */
+class ResetSchedules {
+public:
+    ResetSchedules(Function root) {
+        reset_schedule(root);
+        CallGraph cg(root);
+        vector<Function> calls = cg.transitive_calls(root);
+        for (std::vector<Function>::iterator I = calls.begin(), E = calls.end(); I != E; ++I) {
+            reset_schedule(*I);
+        }
+    }
+    
+private:
+    /** Hackish way of resetting a function schedule to the
+     * default. Mostly yanked from Function::define(). */
+    void reset_schedule(Function f) {
+        Schedule &olds = f.schedule();
+        std::vector<Bound> oldbounds(olds.bounds().begin(), olds.bounds().end());
+        f.schedule() = Schedule();
+        Schedule &s = f.schedule();
```

Comment:
I believe these two lines could be slightly more easily written:

    Schedule &s = f.schedule();
    s = Schedule();